grove/switchboard.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2021 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+Registry for feature toggles used by the Console Backend (Ocim).
+"""
+
+from urllib.parse import urljoin
+
+from django.conf import settings
+from UnleashClient import UnleashClient
+
+SWITCH_GROVE_DEPLOYMENTS: str = "enable_grove_deployments"
+
+# In-memory "cache" for the initialized clients
+__clients__ = dict()
+
+
+def get_unleash_client(url: str, project_id: int, instance_id: str, app_name: str) -> UnleashClient:
+    cache_key = f"{project_id}-{instance_id}-{app_name}"
+
+    initialized_client = __clients__.get(cache_key, None)
+    if initialized_client is not None:
+        return initialized_client
+
+    client = UnleashClient(
+        url=url,
+        instance_id=instance_id,
+        app_name=app_name,
+        custom_headers={
+            'Authorization': settings.GITLAB_API_TOKEN
+        }
+    )
+
+    client.initialize_client()
```

Comment:
@pkulkark @gabor-boros Ocim stage was broken with the log saying that all postgres connections were used up. Restarting Ocim didn't help.

I spent some time debugging that and I think the reason is that initializing the unleash client sets up some periodic tasks in a background thread pool. I'm guessing django waits for all threads scheduled during the course of a request to shut down before releasing database connections, but since the unleash periodic tasks are periodic and kept alive forever, database connections are kept open.

Currently the problem is exacerbated by the fast that attempts to connect to the unleash server fail. I guess some credentials are not set up correctly on stage.  This means that the unleash client never gets cached and a new one is getting set up for every request, scheduling new background tasks, until eventually exhausting the available connection slots on the DB server.

If the unleash credentials were correct, the problem would probably be less noticeable, but it's still not a good idea to call `initialize_client()` inside a request. Ideally we would do that at startup, but I guess we cannot do that since we want to be able to support new repositories dynamically, without having to restart Ocim.

Using the client with a context manager could be a solution, but I'm not sure about performance implications:
```
with UnleashClient(...) as client:
    client.is_enabled(...)
```

I tried that on stage, but then it failed with a different error about a non-existing cache file, which I suppose is related to unleash client failing to connect to the server.

So for now I just changed the `is_feature_enabled` to return `False` on stage and not set up an unleash client.

Ocim stage is now up and running, but the instance list is empty with `NotImplementedError: No serializer enabled for that Instance type.` printed in the log. I guess we're missing a serializer for a GroveInstance.

Reply:
@pkulkark It seems the `No serializer enabled for that Instance type` is not resolved yet. Did you push every commit?

Reply:
@gabor-boros I didn't see the errors while I was testing. Keith mentioned he manually deleted some instances, could that be the reason for this error?

Reply:
I think these are not related, because there were no issues before I created a new instance (by signing up through the registration flow -- properly)